[Improvement] Add hdfs path health check to AppBalanceSelectStorageStrategy

smallzhongfeng opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://github.com/Tencent/Firestorm/blob/master/CONTRIBUTING.md
  2. Ensure you have added or run the appropriate tests for your PR
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]XXXX Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue.
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
The detection logic of abnormal paths is also added to the `APP_BALANCE` strategy.


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Anomaly detection can be performed when selecting a remote path to avoid selecting a problematic path.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released versions or within the unreleased branches such as master.
If no, write 'No'.
-->
The parameters of the original IO_SAMPLE strategy are reused and the names are changed.

1. `rss.coordinator.remote.storage.select.strategy` support `APP_BALANCE` and `IO_SAMPLE`, `APP_BALANCE` selection strategy based on the number of apps, `IO_SAMPLE` selection strategy based on time consumption of reading and writing files.
2. `rss.coordinator.remote.storage.schedule.time` , if user choose `IO_SAMPLE`, file will be read and written at regular intervals.
3. `rss.coordinator.remote.storage.schedule.file.size` , the size of each read / write HDFS file.
4. `rss.coordinator.remote.storage.schedule.access.times`, number of times to read and write HDFS files.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Passed uts.

jerqi commented on coordinator/src/main/java/org/apache/uniffle/coordinator/SelectStorageStrategy.java:
```diff
@@ -17,24 +17,17 @@
+  List<Map.Entry<String, RankValue>> sortPathByRankValue(String path, Path testPath, long time, boolean isHealthy);
 
-  Map<String, RankValue> getRemoteStoragePathRankValue();
+  void setFs(FileSystem fs);
```
We would better not assume our remote storage is only FileSystem.

smallzhongfeng replied:
Could you elaborate more, or give me some other examples ?

jerqi replied:
If we want to support S3 , COS, Redis or some DBs, it's not suitable for us.

smallzhongfeng replied:
Maybe I can remove this method, thinking that if I want to support redis or other databases, I need to re implement the `sortPathByRankValue` method. It's a huge job.

jerqi commented:
Like discussion at dev mail list, we will freeze the code and cut 0.6 version branch in September 15, we  will not merge this pr before I cut 0.6 version branch, are you ok?

jerqi commented:
Like discussion at dev mail list, we will freeze the code and cut 0.6 version branch in September 15, we  will not merge this pr before I cut 0.6 version branch, are you ok?

jerqi commented on coordinator/src/main/java/org/apache/uniffle/coordinator/SelectStorageStrategy.java:
```diff
@@ -17,24 +17,14 @@
-  Map<String, RemoteStorageInfo> getAppIdToRemoteStorageInfo();
-
-  Map<String, RankValue> getRemoteStoragePathRankValue();
+  List<Map.Entry<String, RankValue>> sortPathByRankValue(String path, Path testPath, long time, boolean isHealthy);
```
Path seems to be a concept of filesystem. It may be not general.

smallzhongfeng replied:
Maybe we can support fileSystem at the beginning. If we need to support db, then there may be more things to consider, or we can add a parameter. If it is a file supported by fileSystem, we will turn on this switch.

smallzhongfeng replied:
For `S3` , `COS`, it's OK, right?

jerqi replied:
At least We need consider the object Storage.

smallzhongfeng replied:
IIRC, Object storage such as s3 also has the concept of Path.

smallzhongfeng replied:
But `cos` doesn't seem to have this concept.

zuston replied:
As I know, the object storage also can be accessed by Hadoop Filesystem Interface (HCFS), like s3/oss/cos. So this is not a problem

But if we want to involve more storages which are not implemented by HCFS or speed up specified api which is only compatible with original storage api, maybe we should implement uniffle's own filesystem, it can refer to Flink.

Please let me if I'm wrong.

smallzhongfeng replied:
Thank you very much for your reminder. Although I am not very impressed with this, I have used a similar method to read files in s3.

smallzhongfeng commented:
For object storage, there may be more S3, but for cos, I don't know this very well. Do you have any good suggestions? @jerqi

jerqi commented:
> For object storage, there may be more S3, but for cos, I don't know this very well. Do you have any good suggestions? @jerqi

Do the S3 have the concept of `path`? It only have the concept of bucket.

smallzhongfeng commented:
Could we support anomaly detection of hdfs files first?

smallzhongfeng commented:
Could we support anomaly detection of hdfs files first?

jerqi commented:
> Could we support anomaly detection of hdfs files first?

We should have more general interface. We can only implement part of them.

zuston commented:
> > For object storage, there may be more S3, but for cos, I don't know this very well. Do you have any good suggestions? @jerqi
> 
> Do the S3 have the concept of `path`? It only have the concept of bucket.

Yes, the s3 compatible filesystem of Hadoop has implemented the abstraction of `Path`, but it's slower than the s3 original api when using `list` files/dirs. I think cos is the same with s3.

By the way, I implemented the hadoop filesystem api for internal object store in the past, so have some experience on it.

jerqi commented:
> > > For object storage, there may be more S3, but for cos, I don't know this very well. Do you have any good suggestions? @jerqi
> > 
> > 
> > Do the S3 have the concept of `path`? It only have the concept of bucket.
> 
> Yes, the s3 compatible filesystem of Hadoop has implemented the abstraction of `Path`, but it's slower than the s3 original api when using `list` files/dirs. I think cos is the same with s3.
> 
> By the way, I implemented the hadoop filesystem api for internal object store in the past, so have some experience on it.

We shouldn't relay on the filesystem of object storage too much. Some operations of object storage is slow. For example, `list` operation, `rename` operation and `append` operation.

smallzhongfeng commented:
> We should have more general interface. We can only implement part of them.

I think we only need to provide a path that includes the remote, and the rest of the methods can be implemented by judging different storage systems.

jerqi commented:
> > We should have more general interface. We can only implement part of them.
> 
> I think we only need to provide a path that includes the remote, and the rest of the methods can be implemented by judging different storage systems.

My doubt point is that the concept of path is not general one for the storage.

smallzhongfeng commented:
So I did not use the concept of `path` for the time being, but used a string to represent.

smallzhongfeng commented:
> So I did not use the concept of `path` for the time being, but used a string to represent.
The concept of subsequent use of buckets or paths can be implemented through the readAndWrite interface.

jerqi commented:
> > So I did not use the concept of `path` for the time being, but used a string to represent.
> 
> The concept of subsequent use of buckets or paths can be implemented through the readAndWrite interface.

It's ok.

jerqi commented on coordinator/src/main/java/org/apache/uniffle/coordinator/SelectStorageStrategy.java:
```diff
@@ -17,24 +17,12 @@
-  Map<String, RemoteStorageInfo> getAppIdToRemoteStorageInfo();
-
-  Map<String, RankValue> getRemoteStoragePathRankValue();
+  List<Map.Entry<String, RankValue>> readAndWrite(String path);
```
I have a little doubt, why do the `readAndWrite` become the only interface method for SelectStorageStrategy? I can rank by many methods, it's not necessary to read and write.

smallzhongfeng replied:
The meaning of the `readAndWrite` interface is to perform anomaly detection of remote paths. For hdfs, reading and writing files is a more direct way to detect, so I took this name, but the real comparison method comes from `sortPathByRankValue`, which I do not have. The reason for defining it as an interface is because the objects to be sorted may be different, and this is left to different storage methods for implementation.

jerqi replied:
`readAndWrite` is not a good name. It's strange for the interface `SelectStorageStrategy`.

smallzhongfeng replied:
OK, updated.

jerqi replied:
Should we add a method `pickStorage` or `selectStorage` for this interface?

smallzhongfeng replied:
Yes, that might make more sense for this interface, I will added.

smallzhongfeng commented:
PTAL @jerqi

jerqi commented on coordinator/src/main/java/org/apache/uniffle/coordinator/SelectStorageStrategy.java:
```diff
@@ -17,24 +17,14 @@
 public interface SelectStorageStrategy {
 
-  RemoteStorageInfo pickRemoteStorage(String appId);
+  List<Map.Entry<String, RankValue>> detectStorage(String path);
```
`path -> uri`?
Path is not general concept for storage. Maybe we should use uri.

smallzhongfeng replied:
OK, good suggestion.

jerqi commented on coordinator/src/main/java/org/apache/uniffle/coordinator/SelectStorageStrategy.java:
```diff
@@ -17,24 +17,14 @@
-  Map<String, RemoteStorageInfo> getAppIdToRemoteStorageInfo();
-
-  Map<String, RankValue> getRemoteStoragePathRankValue();
+  List<Map.Entry<String, RankValue>> pickStorage(List<Map.Entry<String, RankValue>> pathList);
```
Could we give `pathList` a better name?

smallzhongfeng replied:
Do you have any good advice?

smallzhongfeng replied:
What about `uriRankings`?

jerqi replied:
`uriList` or `uris` may be better?
Why do the method `pickStorage` return a list?

smallzhongfeng replied:
I modified it to let `pickStorage` return to a selected path.

jerqi commented on coordinator/src/main/java/org/apache/uniffle/coordinator/SelectStorageStrategy.java:
```diff
@@ -17,24 +17,14 @@
-  Map<String, RemoteStorageInfo> getAppIdToRemoteStorageInfo();
-
-  Map<String, RankValue> getRemoteStoragePathRankValue();
+  String pickStorage(List<Map.Entry<String, RankValue>> uris, String appId);
```
Should we return `RemoteStorageInfo`?

smallzhongfeng replied:
No need, because the return value of `pickStorage` only needs to be passed to `incRemoteStorageCounter`, `incRemoteStorageCounter` only needs `String` type, and `String` type is more general than `RemoteStorageInfo`.

jerqi replied:
> No need, because the return value of `pickStorage` only needs to be passed to `incRemoteStorageCounter`, `incRemoteStorageCounter` only needs `String` type, and `String` type is more general than `RemoteStorageInfo`.

`String` is too general to describe what we want to return. `RemoteStorageInfo` is more accurate.

smallzhongfeng replied:
OK, updated.

jerqi commented on coordinator/src/main/java/org/apache/uniffle/coordinator/AppBalanceSelectStorageStrategy.java:
```diff
@@ -39,106 +45,127 @@
+      Path testPath = new Path(rssTest);
+      try {
+        FileSystem fs = HadoopFilesystemProvider.getFilesystem(remotePath, hdfsConf);
+        for (int j = 0; j < readAndWriteTimes; j++) {
```
If we just detect the health of storage, it seems unnecessary for us write and read them multiple times.

smallzhongfeng replied:
For appBalance, it can indeed be changed once, but for io sampling, the number of times can be increased, perhaps we can set the parameter to 1 by default.

jerqi replied:
For appBalance, it shouldn't be a configuration option.

smallzhongfeng replied:
done.

jerqi commented on coordinator/src/main/java/org/apache/uniffle/coordinator/AppBalanceSelectStorageStrategy.java:
```diff
@@ -39,106 +45,127 @@
+  private final Configuration hdfsConf;
+  private final int fileSize;
+  private final int readAndWriteTimes;
+  private boolean remotePathIsHealthy = true;
```
What does this variable `remotePathIsHealthy`  mean?
This class is a strategy. I can't get the meaning of this varibale.

smallzhongfeng replied:
This variable is used to determine whether the path is normal, and to pass the parameter into the method `sortPathByRankValue`.

jerqi replied:
> This variable is used to determine whether the path is normal, and to pass the parameter into the method `sortPathByRankValue`.

One strategy should have multiple paths, every path is healthy, the variable is true?

smallzhongfeng replied:
I know that there are many paths, but this attribute is applied to each path. The healthy path is processed by `sortPathByRankValue` according to `remotePathIsHealthy` as true, and corresponding logic is processed when it is false.

smallzhongfeng replied:
You can look at the logic of `sortPathByRankValue`. When we judge by this parameter, it is a normal path, and then deal with it accordingly.

jerqi replied:
Is the variable necessary to become member field?

smallzhongfeng replied:
Updated.

jerqi commented on coordinator/src/main/java/org/apache/uniffle/coordinator/AppBalanceSelectStorageStrategy.java:
```diff
@@ -39,106 +45,127 @@
+      uris = uris.stream().filter(rv -> rv.getValue().getReadAndWriteTime().get() != Long.MAX_VALUE).sorted(
+          Comparator.comparingInt(entry -> entry.getValue().getAppNum().get())).collect(Collectors.toList());
+    } else {
+      // If all paths are unhealthy, assign paths according to the number of apps
```
Should we add some logs and metrics when all paths are unhealthy?

smallzhongfeng replied:
OK, I will add.

jerqi commented on coordinator/src/main/java/org/apache/uniffle/coordinator/ApplicationManager.java:
```diff
@@ -42,39 +43,59 @@
+        conf.getLong(CoordinatorConf.COORDINATOR_REMOTE_STORAGE_SCHEDULE_TIME), TimeUnit.MILLISECONDS);
+  }
+
+  public void checkReadAndWrite() {
```
Should we put this logic into the class `SelectStorageStrategy`?

jerqi commented:
Maybe we should have a abstract class to define the common behaviors of SelectStorageStrategy.

smallzhongfeng commented:
This modification has done three things

1. An abstract class `AbstractSelectStorageStrategy` is added to provide a method for reading and writing hdfs paths.
2. Added sorted storage list.
3. Removed a global variable of `isHealthy` and changed it to a local variable.

smallzhongfeng commented:
Could you open this workflow for me? @jerqi

smallzhongfeng commented:
PTAL @jerqi

jerqi commented on coordinator/src/main/java/org/apache/uniffle/coordinator/AbstractSelectStorageStrategy.java:
```diff
@@ -0,0 +1,85 @@
+    }
+  }
+
+  @Override
```
If we don't implement the method `detectStorage` and `pickStorage`, we can remove them in the abstract class.

smallzhongfeng replied:
Done.

jerqi reviewed (approved):
LGTM, thanks @smallzhongfeng , wait for CI

jerqi commented:
https://github.com/apache/incubator-uniffle/actions/runs/3213433147 There is a flay test. Could you help me fix it?
[test-reports-spark3.2.0 (1).zip](https://github.com/apache/incubator-uniffle/files/9741113/test-reports-spark3.2.0.1.zip)

smallzhongfeng commented:
> https://github.com/apache/incubator-uniffle/actions/runs/3213433147 There is a flay test. Could you help me fix it? [test-reports-spark3.2.0 (1).zip](https://github.com/apache/incubator-uniffle/files/9741113/test-reports-spark3.2.0.1.zip)

OK, I will fix it.
